### PR TITLE
Update to 1.19.3, add event-based registry inspection and modification API

### PR DESCRIPTION
**Pull request by AlexTMjugador**
A Minecraft version between 1.19.1 and 1.19.3 changed several internal implementation details related to registries and the network codec that broke compatibility with this plugin:

Plugin load exception on 1.19.3

I'm interested in getting this plugin to work on a Paper 1.19.3 server, so I dug into the Minecraft class changes and updated the plugin accordingly. I've tested the changes lightly on a local server by using the example override file mentioned in the README.

### CODE REVIEW

File: nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java [hunk 1]
```diff
@@ -18,22 +18,22 @@ public static void injectCodec(Map<HookType, List<KeyedOverride>> hooks, Logger
...
 
-        Codec<RegistryAccess> CAPTURED_CODEC = RegistryAccess.NETWORK_CODEC;  // Capture the codec stored in the variable, we will be replacing it.
+        Codec<RegistryAccess> CAPTURED_CODEC = RegistrySynchronization.NETWORK_CODEC;  // Capture the codec stored in the variable, we will be replacing it.
 
         // This is meant to basically allow us to encode our own values before it is sent to the client.
         Codec<RegistryAccess> INJECTED_CODEC = new Codec<>() {
 
             @Override
             public <T> DataResult<T> encode(RegistryAccess input, DynamicOps<T> ops, T prefix) {
-                DataResult<JsonElement> result = CAPTURED_CODEC.encode(input, JsonOps.INSTANCE, JsonOps.INSTANCE.empty());
```
Comment (Owen1212055) [inline nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java:28]: Please specify the types rather than using var.

**Review by Owen1212055 — CHANGES_REQUESTED** [review]
Little thing.

Thank you so much for your work! Thanks for using this plugin too. 😄

**Comment by Owen1212055** [comment]
And if you wanna give it a shot, id recommend making registry injection an event rather than having to register it each time.

We can still allow manual registration, but an event in addition would be cool. If you don't want to no worries.

File: nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java (hunk 1 above)
Comment (AlexTMjugador) [reply nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java:28]: Done. Thanks for the prompt review! :heart:

**Comment by AlexTMjugador** [comment]
I would not mind giving that refactor some attention, but I'm not sure how you envision events altering the lifecycle of registry injections, so I don't feel confident doing it yet. Would you mind elaborating on what you want to see? :heart:

**Comment by Owen1212055** [comment]
> I would not mind giving that refactor some attention, but I'm not sure how you envision events altering the lifecycle of registry injections, so I don't feel confident doing it yet. Would you mind elaborating on what you want to see? :heart:

I imagine a sort of RegistryOverrideEvent which will be triggered when the codec runs it's serialization logic. The current "injections" which are provided by the api are a bit limiting, and in general this event would offer as a replacement for those who want more powerful behavior. 

Internally, there will need to be something that listens to this event and uses it for the already existing override system. If that makes sense.

**Comment by AlexTMjugador** [comment]
Thanks for the explanation! It helped me to understand your vision more clearly. I'll give your idea a shot and let you know when I give up or have any results to share :smile:

**Comment by AlexTMjugador** [comment]
I did it! :tada: 

The latest commits add a new `RegistrySendEvent` to the API that is fired by the hooked codec with the registry data that is about to be sent. Listeners can inspect that data and modify it. Any modifications will be applied to the data sent to players unless the event is cancelled. The plugin is now a listener of this event, and uses it to implement the override logic.

File: nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java [hunk 2]
```diff
@@ -34,32 +36,37 @@ public <T> DataResult<T> encode(RegistryAccess input, DynamicOps<T> ops, T prefi
...
+                    JsonArray registry = mainRegistry.get(hookedTypeKey.toString()).getAsJsonObject().getAsJsonArray("value");
 
-                    // Populate it and key it by its identifier, this is because they are stored in an array and this allows us to fetch them nicely.
-                    Map<String, JsonObject> registeredObjects = new HashMap<>(registry.size());
+                    // Populate maps keyed by identifiers, because they are stored in an array and this allows us to fetch them nicely.
+                    Map<NamespacedKey, JsonObject> registryEntries = new HashMap<>(registry.size());
+                    Map<NamespacedKey, JsonObject> registryEntryHolders = new HashMap<>(registry.size());
                     for (JsonElement jsonElement : registry) {
-                        registeredObjects.put(jsonElement.getAsJsonObject().get("name").getAsString(), jsonElement.getAsJsonObject().getAsJsonObject("element"));
+                        var namespacedKey = Objects.requireNonNull(NamespacedKey.fromString(jsonElement.getAsJsonObject().get("name").getAsString()));
```
Comment (Owen1212055) [inline nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java]: var + don't use Objects.requireNonNull.
In general your IDE probably screams at you to add this, but, it's not needed.

File: nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java [hunk 3]
```diff
@@ -34,32 +36,37 @@ public <T> DataResult<T> encode(RegistryAccess input, DynamicOps<T> ops, T prefi
...
-                    // Iterate over the overrides
-                    for (KeyedOverride override : entry.getValue()) {
-                        String overrideKey = override.key().toString();
-                        JsonObject toOverride = registeredObjects.get(overrideKey);
-                        if (toOverride == null) {
-                            LOGGER.warn("Couldn't find override for {} {}", overrideKey, entry.getKey());
-                            continue;
-                        }
+                    // Let event handlers observe and modify the registry data.
+                    final var sendEvent = new RegistrySendEvent(hookedTypeKey, registryEntries);
```
Comment (Owen1212055) [inline nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java]: var

File: nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java [hunk 4]
```diff
@@ -34,32 +36,37 @@ public <T> DataResult<T> encode(RegistryAccess input, DynamicOps<T> ops, T prefi
...
+                    Bukkit.getPluginManager().callEvent(sendEvent);
+
+                    // Apply changes made to the registry entries map only if the registry override event was not cancelled.
+                    if (sendEvent.isCancelled()) {
+                        continue;
+                    }
+
+                    for (final var overriddenEntry : registryEntries.entrySet()) {
```
Comment (Owen1212055) [inline nms/src/main/java/com/owen1212055/biomevisuals/nms/hooks/RegistryHook.java]: var

File: plugin/src/main/java/com/owen1212055/biomevisuals/Main.java [hunk 5]
```diff
@@ -92,6 +98,25 @@ public void onDisable() {
...
     }
 
+    @EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
+    public void onRegistrySend(final @NotNull RegistrySendEvent event) {
+        List<KeyedOverride> keyedOverrides;
+        if ((keyedOverrides = OVERRIDES.get(HookType.getType(event.getRegistryType()))) == null) {
+            return;
+        }
+
+        for (final var keyedOverride : keyedOverrides) {
```
Comment (Owen1212055) [inline plugin/src/main/java/com/owen1212055/biomevisuals/Main.java]: var

**Review by Owen1212055 — CHANGES_REQUESTED** [review]
This is really good!!!
This whole concept is much cleaner in general and will hopefully prove the api more powerful.

One thing that i'm a little bit iffy on is that the underlying JsonObject is exposed. See how we have stuff like BiomeData, it would be nice to instead pass that instead of the raw json.
Generally you're moving the jsonobject "merging" from the internal implementation to the server. 

Perhaps we could have something in the API like a  RegistryKey\<BiomeData\> and RegistryKey\<DimensionData\> (unused).

I donno know, generics with evens like this are a bit iffy, but try playing around.
Overall really happy with this, nice work. 😄

File: plugin/src/main/java/com/owen1212055/biomevisuals/Main.java [hunk 6]
```diff
@@ -105,7 +105,7 @@ public void onRegistrySend(final @NotNull RegistrySendEvent event) {
             return;
         }
 
-        for (final var keyedOverride : keyedOverrides) {
+        for (var keyedOverride : keyedOverrides) {
```
Comment (Owen1212055) [inline plugin/src/main/java/com/owen1212055/biomevisuals/Main.java]: You didn't actually remove the vars. 😢
Just replace it with the actual types.

Comment (AlexTMjugador) [reply plugin/src/main/java/com/owen1212055/biomevisuals/Main.java]: My bad, I thought you wanted me to remove the `final` qualifier only. I'll get rid of `var` now :smile:

**Comment by AlexTMjugador** [comment]
It has been more complicated than I anticipated, but I made the biome registry send event expose strongly-typed `BiomeData` objects instead. The code is designed in such a way that future registry type hooks (i.e., for dimension registries) should get their own registry send event class, so there are no generic type parameters in the event class itself, which I think provides the most comfortable API for end-users.

These strongly-typed objects are now deserialized too, so the `JsonAdapter` class has been moved and expanded. In addition, most registry data holder classes in the API were refactored to records, and their data model streamlined.

This is how using the new event handler API looks like after all this work. I think it's pretty nice, isn't it? :heart:

https://github.com/Owen1212055/BiomeVisuals/blob/99bc84e483af43a261d23eec27d36c7792e803b8/plugin/src/main/java/com/owen1212055/biomevisuals/Main.java#L101-L125

File: nms/src/main/java/com/owen1212055/biomevisuals/nms/ApiEntityConverter.java [hunk 7]
```diff
@@ -1,34 +1,91 @@
...
     }
 
-    public static JsonElement convert(MoodSound moodSound) {
-        AmbientMoodSettings settings = new AmbientMoodSettings(CraftSound.getSoundEffect(moodSound.getSoundEvent()), moodSound.getTickDelay(), moodSound.getBlockSearchExtent(), moodSound.getSoundPositionOffset());
+    public static AmbientParticle deserializeAmbientParticle(JsonElement particle) {
+        AmbientParticleSettings settings = AmbientParticleSettings.CODEC.decode(JsonOps.INSTANCE, particle).map(Pair::getFirst).result().orElseThrow();
+
+        try {
+            // FIXME the particle data object is not serialized as such, so we can't deserialize it back
+            var probabilityField = AmbientParticleSettings.class.getDeclaredField("c");
```
Comment (Owen1212055) [inline nms/src/main/java/com/owen1212055/biomevisuals/nms/ApiEntityConverter.java]: Can you put the Mojang name next to it? And var :(

**Comment by Owen1212055** [comment]
100%! Will look over the weekend, I recommend joining my discord (check the readme)

File: nms/src/main/java/com/owen1212055/biomevisuals/nms/ApiEntityConverter.java (hunk 7 above)
Comment (AlexTMjugador) [reply nms/src/main/java/com/owen1212055/biomevisuals/nms/ApiEntityConverter.java]: Of course, done!

**Review by Owen1212055 — APPROVED** [review]

**Pull request merged**